source/change-streams/change-streams.rst

```diff
@@ -408,7 +416,9 @@ Drivers MUST allow users to specify a ``startAtOperationTime`` option in the ``C
 
 ``startAtOperationTime`` and ``resumeAfter`` are mutually exclusive; if both ``startAtOperationTime`` and ``resumeAfter`` are set, the server will return an error. Drivers MUST NOT throw a custom error, and MUST defer to the server error.
 
-If neither ``startAtOperationTime`` nor ``resumeAfter`` are specified, and the max wire version is >= ``7`` drivers MUST set a default ``startAtOperationTime``. A default timestamp can be obtained from the ``operationTime`` field on any server response (e.g. ``isMaster``). This allows change streams to be resumed before the first notification is received.
+If neither ``startAtOperationTime`` nor ``resumeAfter`` are specified, and the max wire version is >= ``7`` drivers SHOULD set a default ``startAtOperationTime`` using the ``MongoClient`` field ``lastOperationTime`` (see `Causal Consistency <../causal-consistency/causal-consistency.rst#mongoclient-changes>`_.
+
+If the ``lastOperationTime`` is not available, ``startAtOperationTime`` MAY be omitted. If this is the case, then the ``ChangeStream`` MUST save the ``lastOperationTime`` when the initial ``aggregate`` command returns.
```

Comment:
> If the ``lastOperationTime`` is not available, ``startAtOperationTime`` MAY be omitted. If this is the case, then the ``ChangeStream`` MUST save the ``lastOperationTime`` when the initial ``aggregate`` command returns.

Doesn't this change preclude the need to have a `lastOperationTime` in the MongoClient at all? 

Could it not work like: Get last operation time from the ClientSession, if null (i.e. no previous operations on the session) save the initial operation time from the aggregate command in the ChangeStream cursor.  

This allows empty initial batches to be retried with the need to plumb back to the MongoClient.

Reply:
I agree, though is there even any benefit of getting the operation time from the client session?  I mean, 99% of the time it will be an implicit session with no operation time anyway.  I suggest simplifying to just the fallback case.

Reply:
The server design doc specifically said "Start all change streams with a 'startAtOperationTime', and remember that time. ", which is why we do the initial tracking. This change is meant to be a backup in the case that the operationTime is not known.

Reply:
Ok this is actually an interesting question. I think using the operationTime from the aggregate should be sufficient, see below.

When the server receives a change stream without any `resumeAfter` or `startAtOperationTime`, it starts watching at the most-recent *majority committed* oplog entry. In contrast, the operation time returned from each command represents the time of the most recent write, *not necessarily majority-committed*. This means the following thing can happen:
1. A client starts watching a change stream. Suppose the driver does not attach a 'startAtOperationTime', so it starts watching at time _T<sub>majority</sub>_
2. This operation returns with operationTime _T<sub>recent_op</sub>_ > _T<sub>majority</sub>_

There are potentially many oplog entries between _T<sub>recent_op</sub>_ and _T<sub>majority</sub>_. Normally this change stream would return all of them. However, if there's a network error on the next getMore then we will start watching the second time at _T<sub>recent_op</sub>_, and miss everything in between.

Contrast this scenario to if the driver always specified `startAtOperationTIme` when opening a change stream. In that scenario, a network error during the initial aggregate would always return the same results, namely everything after _T<sub>recent_op</sub>_ (or some other operation time the driver has stored, which is probably more recent than _T<sub>majority</sub>_).

Now, it's not clear to me the difference in semantics here matters at all. When you open a new change stream, we don't guarantee much about what you will or will not see. In either case above, if the client opened a change stream, then that client performed a write, it would see that write.

A related case which we were attempting to solve is this: Suppose a client opens a change stream and watches/iterates it for several hours with no results (possibly because of a $match that filtered everything out). After several hours there's a network error during one of the getMores, around the same time that an event happens which the change stream was waiting for. We should not resume that change stream from the most-recent majority-committed oplog entry, because that might be skipping that interesting event. In either proposal here, the driver would correctly restart from the original operationTime (scan through those hours of history 😵) then find the interesting event ✅🙌🎉

The extra hours of scanning is an interesting downside that I don't think we've considered well. In this scenario your oplog probably still has enough history and it'll be fine except maybe bringing a bunch of irrelevant oplog into memory. However, the server will actually error if the time you specified is no longer in the oplog (since we'd potentially miss entries in that case), thus preventing you from resuming in the above scenario if the hours changed to something more like days which was larger than your oplog.

Perhaps the better thing to do here would be for the server to return the information we use internally [in SERVER-29929](https://jira.mongodb.org/browse/SERVER-29929). This information is not returned from normal getMores on change streams; it's only passed between mongos and mongod. We could surface this information though. If we did so, drivers could use this to keep track of the most recent time that the stream is guaranteed not to return anything before. This would avoid scanning hours of oplog and potentially raising an error upon resuming if there was a network error.

Reply:
cc @asya999 and @stbrody

Reply:
You seem to have laid out the behavior and options well, I'll defer to @asya999 and @storch as to what to do from here.

Reply:
Thanks Charlie.  Based on this feedback, we can simplify the spec by requiring use of the operationTime from the aggregate command. Any objections?

Reply:
👍 I'll remove the tracking in the MongoClient and incorporate these changes.

Reply:
@cswanson310 
> In either proposal here, the driver would correctly restart from the original operationTime (scan through those hours of history 😵) 

I think drivers could avoid that by updating the default startAtOperationTime with the operationTime from each (empty) getMore response. Then when the resume happens it resumes from the previous getMore operationTime.

Reply:
@ShaneHarvey I also thought about that - it _seems_ right, but I don't think it is. The operationTime returned on each getMore has no substantial correlation with what the underlying scan on the oplog may or may not have consumed yet. This is why we had to invent a new thing for SERVER-29929.

In particular, the collection scan on the oplog is going to stop at the most-recent majority-committed entry, but the operationTime could still be ahead of that. Anything in between those two could be lost.

@jyemin that sounds reasonable to me!

Reply:
Ah thanks for explaining that again Charlie! I missed it in your original comment.